05-type-classes/hw-05.hs

```diff
@@ -0,0 +1,254 @@
+    add (MinMax firstNum) (MinMax secondNum) = MinMax $ max firstNum secondNum
+
+instance Expr Mod7 where
+    lit num                              = Mod7 $ num `mod` 7
+    mul (Mod7 firstNum) (Mod7 secondNum) = Mod7 $ (firstNum + secondNum) `mod` 7
+    add (Mod7 firstNum) (Mod7 secondNum) = Mod7 $ (firstNum * secondNum) `mod` 7
+
+
+testExp :: Expr a => Maybe a
+testExp      = parseExp lit add mul "(3 * -4) + 5"
+testInteger  = testExp :: Maybe Integer
+testBool     = testExp :: Maybe Bool
+testMM       = testExp :: Maybe MinMax
+testSat      = testExp :: Maybe Mod7
+
+
+{-
+Exercise 5 (do this OR exercise 6)
+The folks down in hardware have finished our new custom CPU,
+so we’d like to target that from now on. The catch is that a stack- based architecture was chosen to save money. You need to write a
+cis 194: homework 5 4
+version of your calculator that will emit assembly language for the new processor.
+The hardware group has provided you with StackVM.hs, which is a software simulation of the custom CPU. The CPU supports six operations, as embodied in the StackExp data type:
+ data StackExp = PushI Integer
+               | PushB Bool
+               | Add
+               | Mul
+               | And
+               | Or
+                 deriving Show
+ type Program = [StackExp]
+PushI and PushB push values onto the top of the stack, which can store both Integer and Bool values. Add, Mul, And, and Or each pop the top two items off the top of the stack, perform the appropriate operation, and push the result back onto the top of the stack. For example, executing the program
+ [PushB True, PushI 3, PushI 6, Mul]
+will result in a stack holding True on the bottom, and 18 on top of that.
+If there are not enough operands on top of the stack, or if an op- eration is performed on operands of the wrong type, the processor will melt into a puddle of silicon goo. For a more precise specifica- tion of the capabilities and behavior of the custom CPU, consult the reference implementation provided in StackVM.hs.
+Your task is to implement a compiler for arithmetic expressions. Simply create an instance of the Expr type class for Program, so that arithmetic expressions can be interpreted as compiled programs. For any arithmetic expression exp :: Expr a => a it should be the case that
+ stackVM exp == Right [IVal exp]
+Note that in order to make an instance for Program (which is a type synonym) you will need to enable the TypeSynonymInstances language extension, which you can do by adding
+{-# LANGUAGE TypeSynonymInstances #-}
+as the first line in your file.
+Finally, put together the pieces you have to create a function
+cis 194: homework 5 5
+
+compile :: String -> Maybe Program
+which takes Strings representing arithmetic expressions and com-
+piles them into programs that can be run on the custom CPU.
+-}
+--Exercise 5
+instance Expr StackVM.Program where
+    lit num = [StackVM.PushI num]
```

Comment:
This implementation is correct, BUT (what follows is optional, quite out of scope of this homework, but it's good to know an issue like this exists :smile: ) it will suffer from quadratic complexity due to usage of list concatenation. The reason is similar to why "naive list reverse implementation" is bad (see this SO answer https://stackoverflow.com/questions/26847192/reverse-a-list-in-haskell#answer-56650800)

There is a simple technique to work around this, which it to use difference lists.
The idea is briefly and beautifully explained [here](https://www.youtube.com/watch?v=5-P0Jjku3cY&feature=youtu.be&t=1726) by Joachim Breitner (just need to watch 2 minutes starting from the timestamp in url).

Reply:
@jhrcek Thank you for pointing this out 👍, really appreciate it.

Reply:
Yes! It's mind blowing how simple but effective this solution is. A Professor had once shared this paper with me (which I presume is the original solution, at least in published form): https://www.cs.tufts.edu/comp/150FP/archive/john-hughes/lists.pdf